Mathlib/LinearAlgebra/Matrix/HermitianFunctionalCalculus.lean

```diff
@@ -0,0 +1,302 @@
+
+/-- An n x n matrix over a ring has a finite spectrum. -/
+theorem Matrix.finite_spectrum (A : Matrix n n R) : Set.Finite (spectrum R A) := by
+  rw [← AlgEquiv.spectrum_eq (Matrix.toLinAlgEquiv <| Pi.basisFun R n) A]
+  exact Module.End.finite_spectrum _
+
+instance Matrix.instFiniteSpectrum (A : Matrix n n R) : Finite (spectrum R A) :=
+  Set.finite_coe_iff.mpr (Matrix.finite_spectrum A)
+
+end FiniteSpectrum
+
+section SpectrumDiagonal
+
+variable {R : Type*} [Field R] {n : Type*} [DecidableEq n][Fintype n]
+
+open Module.End
+
+/--  Standard basis vectors are eigenvectors of any associated diagonal linear operator. -/
+lemma Matrix.hasEigenvector_toLin'_diagonal (d : n → R) (i : n) :
+    Module.End.HasEigenvector (Matrix.toLin' (diagonal d)) (d i) (Pi.basisFun R n i) := by
+  constructor
+  · rw [mem_eigenspace_iff]
+    ext j
+    simp only [diagonal, Pi.basisFun_apply, toLin'_apply, mulVec_stdBasis_apply, transpose_apply,
+      of_apply, Pi.smul_apply, LinearMap.stdBasis_apply', smul_eq_mul, mul_ite, mul_one, mul_zero]
+    split_ifs
+    all_goals simp_all
+  · rw [Function.ne_iff]; simp
+
+/-- Eigenvalues of a diagonal linear operator are the diagonal entries. -/
+lemma Matrix.hasEigenvalue_toLin'_diagonal_iff (d : n → R) {μ : R} :
+    HasEigenvalue (toLin' (diagonal d)) μ ↔ ∃ i, d i = μ := by
+  have (i : n) : HasEigenvalue (toLin' (diagonal d)) (d i) := by
+    exact hasEigenvalue_of_hasEigenvector <| hasEigenvector_toLin'_diagonal d i
+  constructor
+  · contrapose!
+    intro hμ h_eig
+    have h_iSup : ⨆ μ ∈ Set.range d, eigenspace (toLin' (diagonal d)) μ = ⊤ := by
+      rw [eq_top_iff, ← (Pi.basisFun R n).span_eq, Submodule.span_le]
+      rintro - ⟨i, rfl⟩
+      simp only [SetLike.mem_coe]
+      apply Submodule.mem_iSup_of_mem (d i)
+      apply Submodule.mem_iSup_of_mem ⟨i, rfl⟩
+      rw [mem_eigenspace_iff]
+      exact (hasEigenvector_toLin'_diagonal d i).apply_eq_smul
+    have hμ_not_mem : μ ∉ Set.range d := by simpa using fun i ↦ (hμ i)
+    have := eigenspaces_independent (toLin' (diagonal d)) |>.disjoint_biSup hμ_not_mem
+    rw [h_iSup, disjoint_top] at this
+    exact h_eig this
+  · rintro ⟨i, rfl⟩
+    exact this i
+
+/-- The spectrum of the diagonal operator is the range of the diagonal viewed as a function. -/
+lemma Matrix.spectrum_diagonal (d : n → R) :
+    spectrum R (diagonal d) = Set.range d := by
+  ext μ
+  rw [← AlgEquiv.spectrum_eq (Matrix.toLinAlgEquiv <| Pi.basisFun R n),
+    ← hasEigenvalue_iff_mem_spectrum, Set.mem_range]
+  exact Matrix.hasEigenvalue_toLin'_diagonal_iff d
+
+end SpectrumDiagonal
+namespace Matrix
+
+variable {𝕜 : Type*} [RCLike 𝕜] {n : Type*} [Fintype n]
+
+namespace IsHermitian
+
+variable [DecidableEq n]
+
+variable {A : Matrix n n 𝕜} (hA : IsHermitian A)
+
+/--Eigenvalues of a Hermitian Matrix, coerced, belong to the spectrum of the assoc.toEuclideanLin -/
+theorem eigenvalue_mem_toEuclideanLin_spectrum_RCLike (i : n) :
```

Comment:
Some of these spectrum/eigenvalue results can probably go in the `Matrix.IsHermitian` file.

Reply:
Should I do this as part of the main PR? If so, I'd like to know why you think so, i.e. why is this different than the above sections you suggested that I PR separately?

Reply:
I was only able to move one of these, as the others require `spectrum` so would need an import. What do you suggest?